tests/test_decoder.py

```diff
@@ -0,0 +1,226 @@
+
+        self.assertTrue(uses_bom('utf_16'))
+        self.assertFalse(uses_bom('utf_16_be'))
+        self.assertTrue(uses_bom('utf_16_be_sig'))
+        self.assertFalse(uses_bom('utf_16_le'))
+        self.assertTrue(uses_bom('utf_16_le_sig'))
+
+        self.assertFalse(uses_bom('utf_8'))
+        self.assertTrue(uses_bom('utf_8_sig'))
+
+    def test_encoding_detectable(self):
+        # No BOM + nothing
+        self.assertFalse(encoding_detectable('', 'utf_32_be'))
+        self.assertFalse(encoding_detectable('', 'utf_32_le'))
+
+        self.assertFalse(encoding_detectable('', 'utf_16_be'))
+        self.assertFalse(encoding_detectable('', 'utf_16_le'))
+
+        self.assertFalse(encoding_detectable('', 'utf_8'))
+        # BOM + nothing
+        self.assertTrue(encoding_detectable('', 'utf_32'))
+        self.assertTrue(encoding_detectable('', 'utf_32_be_sig'))
+        self.assertTrue(encoding_detectable('', 'utf_32_le_sig'))
+
+        self.assertTrue(encoding_detectable('', 'utf_16'))
+        self.assertTrue(encoding_detectable('', 'utf_16_be_sig'))
+        self.assertTrue(encoding_detectable('', 'utf_16_le_sig'))
+
+        self.assertTrue(encoding_detectable('', 'utf_8_sig'))
+        # No BOM + non-ASCII
+        self.assertFalse(encoding_detectable('Ⓝⓔ', 'utf_32_be'))
+        self.assertFalse(encoding_detectable('ⓥⓔ', 'utf_32_le'))
+
+        self.assertFalse(encoding_detectable('ⓡ ', 'utf_16_be'))
+        self.assertFalse(encoding_detectable('ⓖⓞ', 'utf_16_le'))
+
+        self.assertFalse(encoding_detectable('ⓝⓝ', 'utf_8'))
+        # No BOM + ASCII
+        self.assertTrue(encoding_detectable('a ', 'utf_32_be'))
+        self.assertTrue(encoding_detectable('gi', 'utf_32_le'))
+
+        self.assertTrue(encoding_detectable('ve', 'utf_16_be'))
+        self.assertTrue(encoding_detectable(' y', 'utf_16_le'))
+
+        self.assertTrue(encoding_detectable('ou', 'utf_8'))
+        # BOM + non-ASCII
+        self.assertTrue(encoding_detectable('␣ⓤ', 'utf_32'))
+        self.assertTrue(encoding_detectable('ⓟ␤', 'utf_32_be_sig'))
+        self.assertTrue(encoding_detectable('Ⓝⓔ', 'utf_32_le_sig'))
+
+        self.assertTrue(encoding_detectable('ⓥⓔ', 'utf_16'))
+        self.assertTrue(encoding_detectable('ⓡ␣', 'utf_16_be_sig'))
+        self.assertTrue(encoding_detectable('ⓖⓞ', 'utf_16_le_sig'))
+
+        self.assertTrue(encoding_detectable('ⓝⓝ', 'utf_8_sig'))
+        # BOM + ASCII
+        self.assertTrue(encoding_detectable('a ', 'utf_32'))
+        self.assertTrue(encoding_detectable('le', 'utf_32_be_sig'))
+        self.assertTrue(encoding_detectable('t ', 'utf_32_le_sig'))
+
+        self.assertTrue(encoding_detectable('yo', 'utf_16'))
+        self.assertTrue(encoding_detectable('u ', 'utf_16_be_sig'))
+        self.assertTrue(encoding_detectable('do', 'utf_16_le_sig'))
+
+        self.assertTrue(encoding_detectable('wn', 'utf_8_sig'))
+
+
+class DecoderTestCase(unittest.TestCase):
```

Comment:
In complement to testing various encoded `test_strings`, could you add a bunch of byte arrays (e.g. `b'\xc3\xa7a va ?'`, `b'\xef\xbb\xbf\xc3\xa7a va ?'`, `b'\xe7\x00a\x00 \x00v\x00a\x00 \x00?\x00'` etc.) hardcoded inside our test files?

Reply:
Done.

Reply:
That's great!
A single ASCII character, a string that starts with an ASCII character, a single non-ASCII character, a string that starts with a non-ASCII character: I don't see what else we could test.